keps/sig-api-machinery/2896-openapi-v3/README.md

```diff
@@ -0,0 +1,524 @@
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+For alpha, unit tests will be included to ensure that v3 schemas are properly generated and published. A validator will also be used to ensure that the spec generated is valid OpenAPI v3.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Native types are updated to capture capabilities introduced with v3
+- Definition names of native resources are updated to omit their package paths
+- Parameters are reused as components
+- `kubectl explain` to support using the OpenAPI V3 Schema
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+- Enabling/disabling the feature gate
+- For upgrading, aggregated apiserver images must also be updated to serve OpenAPI v3 in order for the aggregator to pick up the spec and publish OpenAPI v3. OpenAPI v2 will continue to work regardless of version skew.
+- For downgrading, aggregated apiservers do not need to be downgraded as everything will revert to using OpenAPI v2 and the OpenAPI v3 endpoint will be untouched.
+
+### Version Skew Strategy
+
+<!--
+If applicable, how will the component handle version skew with other
+components? What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+There is a possible skew between kube-apiserver and old aggregated apiservers. The affected aggregated apiservers will not publish v3 and kube-apiserver will not publish the v3 schema for the group-version claimed by the apiservers until they are upgraded to v3.
+
+## Production Readiness Review Questionnaire
```

Comment:
please include the entire template and just leave the answers for the beta and stable questions blank.

Reply:
updated